### PR TITLE
Support field reference in `Value`

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR extends `Value` type with the ability to carry a reference to a value in another field. 

One of the usage scenarios for such a feature is validation options using references to another field. For example, consider the following proto type `Range` in which the `min_value` can have a value not bigger than `max_value`.
```proto
message Range {
    int32 min_value = 1 [(max).value = "max_value"];
    int32 max_value = 2;
}
```
A reference could made to a nested field using dot-separated fields names in the path.

## Changes to `Value` and related code
 * The type `Reference` was introduced.
 * The type `Value` got the new field `reference`.
 * `MinOption` and `MaxOption` got extensions `parse()` for obtaining corresponding instances of `Value`.

## Other notable changes
 * The event `FieldOptionDiscovered` got a new field called `subject` with type `Field`. The new field is aimed to replace now deprecated fields `type` and `field`.
 * The creation of a singular value via `FieldDescriptor` made stricter, failing with `IllegalStateException` when unsupported `JavaType` enum item is encountered. This is an unlikely case, but it's still better to throw rather than silently return `NULL`.
 * `ProtobufSourceFile` type got extension for finding a message type by its simple name and possible enclosing type names.
* `TypeSystem` got an extension `resolve()` which allows to obtain a `Field` instance for a field of a message type specified by `FieldPath`.

### CODE REVIEW

File: api/src/main/kotlin/io/spine/protodata/value/Options.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                    type = field.type
+                    target = ensureField(FieldPath(value))
+                }
+            }
+            else -> error("${optionPath()} has the value with unexpected format (`$value`).")
+        }
+    }
+
+    private fun optionPath() =
+        "The `($optionName)` option declared in the field `${field.name.value}` of" +
```
Comment (armiol) [inline api/src/main/kotlin/io/spine/protodata/value/Options.kt:91]: Maybe, "for the field"?

**Review by armiol — APPROVED** [review]
@alexander-yevsyukov LGTM with a minor comment to consider.

File: api/src/main/kotlin/io/spine/protodata/value/Options.kt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * Parses the value of this `(min)` option into an instance of [Value].
+ *
+ * The value could be an integer, double, or a reference to another field.
+ */
+public fun MinOption.parse(field: Field, typeSystem: TypeSystem): Value =
+    OptionValue("min", value, field, typeSystem).parse()
+
+/**
+ * Parses the value of this `(min)` option into an instance of [Value].
```
Comment (yevhenii-nadtochii) [inline api/src/main/kotlin/io/spine/protodata/value/Options.kt:51]:
```suggestion
 * Parses the value of this `(max)` option into an instance of [Value].
```

File: api/src/main/proto/spine/protodata/value.proto [hunk 3]
```diff
@@ -166,3 +171,19 @@ message MapValue {
...
+
+// A reference to a value in the field specified in the [target] property.
+//
+// Typically such a reference is obtained by parsing an option of a "source" field.
+//
+message Reference {
+
+    // The type of the referenced value.
+    //
+    // If the reference is obtained by parsing a "source" field option, it is the type of the field.
```
Comment (yevhenii-nadtochii) [inline api/src/main/proto/spine/protodata/value.proto:183]: I don't fully get this sentence.

Did you mean "If ..., it is the type of the **referenced** field."? But then what if not?

File: java/src/test/kotlin/io/spine/protodata/java/MessageTypeExtsSpec.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import io.spine.protodata.protobuf.toPbSourceFile
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+
+@DisplayName("`MessageType` extensions for Java should")
+internal class MessageTypeExtsSpec {
+
+    @Nested inner class
+    `find its class` {
```
Comment (yevhenii-nadtochii) [inline java/src/test/kotlin/io/spine/protodata/java/MessageTypeExtsSpec.kt:47]: We don't need a nested class here. The second test case is negative.

**Review by yevhenii-nadtochii — APPROVED** [review]
LGTM with a couple suggestions.

**Review by armiol — APPROVED** [review]

**Pull request merged**